Document ibc (#884)
CosmWasm/cosmwasm: IBC.md

# IBC interfaces for CosmWasm contracts

If you import `cosmwasm-std` with the `stargate` feature flag, it will expose a
number of IBC-related functionality. This requires that the host chain is

> running an IBC-enabled version of `wasmd`, that is `v0.16.0` or higher. You will

Reviewer: Do I understand correctly that other projects use the "x/wasm" module, and not the wasmd node application?

```suggestion
running an IBC-enabled version of `x/wasm`, that is `v0.16.0` or higher. You will
```

your tokens back.

## Writing New Protocols

However, we go beyond simply _using_ existing IBC protocols, and allow you to
_implement_ your own ICS protocols inside the contract. A good example to
understand this is the
[`cw20-ics20` contract](https://github.com/CosmWasm/cosmwasm-plus/tree/v0.6.0-beta1/contracts/cw20-ics20)
included in the `cosmwasm-plus` repo. This contract speaks the `ics20-1`
protocol to an external blockchain just as if it were the `ibctransfer` module
in Go. However, we can implement any logic we want there and even hot-load it on
a running blockchain.

This particular contract above accepts
[cw20 tokens](https://github.com/CosmWasm/cosmwasm-plus/tree/v0.6.0-beta1/packages/cw20)
and sends those to a remote chain, as well as receiving the tokens back and
releasing the original cw20 token to a new owner. It does not (yet) allow
minting coins originating from the remote chain. I recommend opening up the
source code for that contract and refering to it when you want a concrete
example for anything discussed below.

In order to enable IBC communication, a contract must expose the following 6
entry points. Upon detecting such an "IBC-Enabled" contract, the

> [wasmd runtime](https://github.com/CosmWasm/wasmd) will automatically bind a

Reviewer: ```suggestion
[x/wasm runtime](https://github.com/CosmWasm/wasmd) will automatically bind a
```

on the ordering and version in the `IbcChannel` information, but you could
enforce other constraints as well:

```rust
#[entry_point]
/// enforces ordering and versioning constraints
pub fn ibc_channel_open(deps: DepsMut, env: Env, channel: IbcChannel) -> StdResult<()> { }
```

This is the
[IbcChannel structure](https://github.com/CosmWasm/cosmwasm/blob/v0.14.0-beta4/packages/std/src/ibc.rs#L70-L81)
used heavily in the handshake process:

```rust
#[derive(Serialize, Deserialize, Clone, Debug, PartialEq, JsonSchema)]
pub struct IbcChannel {
    pub endpoint: IbcEndpoint,
    pub counterparty_endpoint: IbcEndpoint,
    pub order: IbcOrder,
    pub version: String,
    /// CounterpartyVersion can be None when not known this context, yet
    pub counterparty_version: Option<String>,
    /// The connection upon which this channel was created. If this is a multi-hop
    /// channel, we only expose the first hop.
    pub connection_id: String,
}

#[derive(Serialize, Deserialize, Clone, Debug, PartialEq, JsonSchema)]
pub struct IbcEndpoint {
    pub port_id: String,
    pub channel_id: String,
}
```

Note that neither `counterparty_version` nor `counterparty_endpoint` is set in

> `ibc_channel_open` for chain A. Chain B should enforce any

Reviewer: How is this possible with non-optional type `counterparty_endpoint: IbcEndpoint`?

Author: I think it just gets "" fields. You are right, we should make this optional if it is not present, but maybe that would confused all calls but one where it is optional (I can use `IbcChannelWithOptionalCounterparty` maybe for this?)

Author: We pass in the counterparty endpoint on init as well: https://github.com/CosmWasm/wasmd/blob/master/x/wasm/ibc.go#L48

I am not sure if these are set in the sdk before calling us

Author: After digging more in the sdk code, I found that counterparty.port is set here (in `ChannelOpenInit`), but counterparty.channel is not set.

https://github.com/cosmos/cosmos-sdk/blob/88e03e4f4056e8614847fd6adf17ce9252b6a89b/x/ibc/core/04-channel/types/msgs.go#L41-L60

In `ChannelOpenTry`, the counterparty channel must be set (ignore that comment, the code shows otherwise): https://github.com/cosmos/cosmos-sdk/blob/88e03e4f4056e8614847fd6adf17ce9252b6a89b/x/ibc/core/04-channel/types/msgs.go#L131-L134

Since we collapse these 2 into one call (to limit the number of exports and duplicated code), I am not sure the best way to handle this.

Making `channel` optional in all `IbcEndpoint` would be overkill and complicate other logic. Using `counterparty: IbcEndpointWithOptionalChannel` for a custom type for `ibc_channel_open` could work, along with a comment that `counterparty_version` and `counterparty.channel` are both set or both unset.

Any other ideas?

Reviewer: Do we need `counterparty_endpoint.port_id` when `counterparty_version` and `counterparty_version.channel` is unset? If not, I think

```
pub counterparty_endpoint: Option<IbcEndpoint>,
```

would be solid.

Author: Some protocols (like ics27 currently) try to enforce the remote port.
Few if any care about the remote channel_id

Reviewer: Hmm, okay. Then I don't have a great idea. In this case it starts feeling wrong to use the same type for two different calls because we don't have dynamic typing available here.

been approved by the remote side.

#### Channel Connect

Once both sides have returned `Ok()` to `ibc_channel_open`, we move onto the
second step of the handshake, which is equivalent to `ChanOpenAck` and
`ChanOpenConfirm` from the spec:

```rust
#[entry_point]
/// once it's established, we may take some setup action
pub fn ibc_channel_connect(
    deps: DepsMut,
    env: Env,
    channel: IbcChannel,
) -> StdResult<IbcBasicResponse> { }
```

At this point, it is expected that the contract updates its internal state and
may return `CosmosMsg` in the `Reponse` to interact with other contracts, just
like in `execute`. In particular, you will most likely want to store the local
channel_id (`channel.endpoint.channel_id`) in the contract's storage, so it
knows what open channels it has (and can expose those via queries or maintain
state for each one).

Once this has been called, you may expect to send and receive any number of
packets with the contract. The packets will only stop once the channel is closed
(which may never happen).

### Channel Close

A contract may request to close a channel that belongs to it via the following

> `CosmosMsg`:

Reviewer: ```suggestion
`CosmosMsg::Ibc`:
```

```rust
pub enum IbcMsg {
    /// This will close an existing channel that is owned by this contract.

>     /// Port is auto-assigned to the contracts' ibc port

Reviewer: ```suggestion
    /// Port is auto-assigned to the contract's IBC port
```

    CloseChannel { channel_id: String },
}
```

Once a channel is closed, whether due to an IBC error, at our request, or at the
request of the other side, the following callback is made on the contract, which
allows it to take appropriate cleanup action:

```rust
#[entry_point]
pub fn ibc_channel_close(
    deps: DepsMut,
    env: Env,
    channel: IbcChannel,
) -> StdResult<IbcBasicResponse> { }
```

### Packet Lifecycle

Unfortunately the
[IBC spec on Pakcet Lifecycle](https://github.com/cosmos/ibc/tree/master/spec/core/ics-004-channel-and-packet-semantics#packet-flow--handling)
is missing all useful diagrams, but it may provide some theoretical background
for this text if you wish to look.

In short, IBC allows us to send packets from chain A to chain B and get a
response from them. The first step is the contract/module in chain A requesting
to send a packet. This is then relayed to chain B, where it "receives" the
packet and calculates an "acknowledgement" (which may contain a success result

> or an error message, as opaque bytes to be interpretted by the sending

Reviewer: ```suggestion
or an error message, as opaque bytes to be interpreted by the sending
```

contract). The acknowledgement is then relayed back to chain A, completing the
cycle.

In some cases, the packet may never be delivered, and if it is proven not to be

> delivered before the timeout period, this can abort the packet, calling the

Reviewer: A timeput is not a period, right? It's a point in time.

```suggestion
delivered before the timeout, this can abort the packet, calling the
```

"timeout". No "receive" nor "acknowledgement" callbacks are ever executed.

#### Sending a Packet

In order to send a packet, a contract can simply return `IbcMsg::SendPacket`
along with the channel over which to send the packet (which you saved in
`ibc_channel_connect`), as well as opaque data bytes to be interpreted by the
other side. You must also return a timeout either as block height or block time
of the remote chain, just like in the ICS20 `Transfer` messages above:

```rust
pub enum IbcMsg {
    /// Sends an IBC packet with given data over the existing channel.
    /// Data should be encoded in a format defined by the channel version,
    /// and the module on the other side should know how to parse this.
    SendPacket {
        channel_id: String,
        data: Binary,
        /// block height after which the packet times out.
        /// at least one of timeout_block, timeout_timestamp is required
        timeout_block: Option<IbcTimeoutBlock>,
        /// block timestamp (nanoseconds since UNIX epoch) after which the packet times out.
        /// See https://golang.org/pkg/time/#Time.UnixNano
        /// at least one of timeout_block, timeout_timestamp is required

>         timeout_timestamp: Option<u64>,

Reviewer: Random idea: We could make this type save (only allow 3 of the 4 states) by creating a dedicated IbcTimeout enum with 3 cases (timestamp, height, both).

Author: This is a bit more annoying from the Go side of things, but maybe it does make sense.
When I thought it was time or block, I was going to use an enum. With all 3 I figured that was a pain.

Author: ```rust
pub enum IbcTimeout {
    Timestamp(u64),
    Block(IbcTimeoutBlock),
    Both{
        timestamp: u64,
        block: IbcTimeoutBlock,
    }
}
```

Would give us eg. `timeout: IbcTimeout::Timestamp(1234)` instead of `timeout_block: None, timeout_timestamp: Some(1234)` and makes sense for `IbcTimeout::Block` as well. `IbcTimeout::Both` would be rather verbose...

```
timeout: IbcTimeout::Both{
  timestamp: 1234,
  block: IbcTimeoutBlock { revision: 1, height: 5000 },
}
```

vs

```
timeout_timestamp: Some(1234),
timeout_block: Some({ revision: 1, height: 5000 }),
```

on reflection, I guess this is not much worse in this case for rust, and cleaner for the others. The Go complexity we just code once (and not every contract), so, yeah I will go with this suggestion.

Reviewer: They can all be positional:

```rust
pub enum IbcTimeout {
    Timestamp(u64),
    Block(IbcTimeoutBlock),
    Both(u64, IbcTimeoutBlock),
}
```

Author: I made this adjustment in  f13fc2e  and  298a421 happy for feedback on that.

Author: > They can all be positional:

Mixed positionals are very difficult to parse in Go (or many other languages) - mixed type arrays are not friendly in most languages. I would avoid that.

Reviewer: Ahh, I see. Makes sense, good point.

```rust
#[entry_point]
pub fn ibc_packet_receive(
    deps: DepsMut,
    env: Env,
    packet: IbcPacket,
) -> StdResult<IbcReceiveResponse> { }
```

Note the different return response here (`IbcReceiveResponse` rather than
`IbcBasicResponse`)? This is because it has an extra field
`acknowledgement: Binary`, which must be filled out. That is the response bytes
that will be returned to the original contract, informing it of failure or
success. (Note: this is vague as it will be refined in the next PR)

Here is the
[`IbcPacket` structure](https://github.com/CosmWasm/cosmwasm/blob/v0.14.0-beta4/packages/std/src/ibc.rs#L129-L146)
that contains all information needed to process the receipt. You can generally
ignore timeout (this is only called if it hasn't yet timed out) and sequence
(which is used by the IBC framework to avoid duplicates). I generally use
`dest.channel_id` like `info.sender` to authenticate the packet, and parse
`data` into a `PacketMsg` structure, using the same encoding rules as we
discussed in the last section.

After that you can process `PacketMsg` more or less like an `ExecuteMsg`,
including calling into other contracts. The only major difference is that you

> must return Acknowledgement bytes in the protocol-specified format

Reviewer: ```suggestion
must return Acknowledgement bytes in the protocol-specified format.
```

returned as part of the acknowledgement (like
[storing the remote address after calling "WhoAmI"](https://github.com/CosmWasm/cosmwasm/blob/v0.14.0-beta4/contracts/ibc-reflect-send/src/ibc.rs#L157-L192)
in our simple `ibc-reflect` example.

On error, you will want to revert any state that was pending based on the
packet. For example, in ics20, if the
[remote chain rejects the packet](https://github.com/CosmWasm/cosmwasm-plus/blob/v0.6.0-beta1/contracts/cw20-ics20/src/ibc.rs#L246),
we must
[return the funds to the original sender](https://github.com/CosmWasm/cosmwasm-plus/blob/v0.6.0-beta1/contracts/cw20-ics20/src/ibc.rs#L291-L317).

#### Handling Timeouts

If the packet was not received on chain B before the timeout, we can be certain
that it will never be processed there. In such a case, a relayer can return a
timeout proof to cancel the pending packet. In such a case the calling contract
will never get `ibc_packet_ack`, but rather `ibc_packet_timeout`. One of the two
calls will eventually get called for each packet that is sent as long as there
is a functioning relayer. (In the absence of a functioning relayer, it will

> never get a response).

Reviewer: ‼️‼️‼️‼️‼️ 😄
